hips/hip-9999.md

```diff
@@ -0,0 +1,96 @@
+
+The HIP also introduces the ability for SDK users to use their own logging library with the SDK. Users will be able to provide a logger implementing `slog.Handler` interface, allowing them to consume instrumented logs in the SDK.
+
+## Motivation
+
+The current logging mechanism in Helm CLI and SDK lacks granularity and structure. Users and developers face difficulties when:
+
+- Troubleshooting Helm CLI commands due to inability to increase the details that can get logged. More logs come in handy when submitting bug reports for example.
+- Instrumenting Helm code with clear and consistent logging conventions for contributors, as well as consumers of CLI and SDK logs.
+- Inflexibility using a different logging library to what the SDK uses. There is no way, in code, to have client loggers consume logs produced by the SDK
+
+The proposed solution addresses these gaps by integrating a logging library that supports log levels and structured output, as well as a widely supported logger implementation interface for SDK users to consume.
+
+## Rationale
+
+The following libraries were evaluated:
+
+1. **`slog`:**
+   - Provides log levels (`info`, `debug`, `error`, `warning`). More levels can be added
+   - Offers structured logging.
+   - Lightweight and straightforward to integrate.
+   - Is part of Go standard library
+   - Has `slog.Handler` interface that other logger implementations can implement to be `slog` compatible
+
+2. **`klog`:**
+   - Provides log levels and also verbosity levels, enabling more granular control. Verbosity level controls the amount of detail in logs, enabling users to filter messages based on their importance or relevance.
+   - Allows getting instrumented Kubernetes client library
+   - Its an external library, hence introducing a dependency
+
+3. **`logr`**
+   - Library defining interfaces for logger implementations and libraries expecting loggers that implement these APIs. `klog`, `slog`, `log` and many others are compatible with `logr`.
+   - Provides log levels also verbosity levels
+   - Its an external library, hence introducing a dependency
+   - Has `logr.Handler` interface that other logger implementations can implement to be `logr` compatible
+
+While these libraries meet all the requirements we want, `slog` is the preferred choice here. It provides all functionality that the other choices have but stands out since it has a wider adoption given that its part of Go standard library. This will be benefitial to SDK users who have their existing logger implementations present. They would either implement `slog.Handler` interface or add a dependency of an existing adapter logger implementation.
+
+## Specification
+
+- Helm will instrument `debug`, `info`, `warning` and `error` log levels. These log levels will enable filtering of logs.
```

Comment:
"fatal" as well?

Reply:
"fatal" would be a custom `slog` log level. This means needing an custom handler. SDK users would not benefit from this however because of
- If they provide their own logger implementing `slog.Handler`, the new log level would no longer have meaning
- Assuming "fatal" log level is meant to "slog.Error(msg); os.Exit(1)", we would not use this in the SDK anyway.

This leave "fatal" log level to just the CLI. It would be easier to just have the code below for the CLI implementation to call

```go
func fatal(msg string, args ...any) {
	slog.Error(msg, args...)
	os.Exit(1)
}
```

Given the choice of `slog`, we will be limited to using the log levels it defines. The same limitation would be valid even if we chose `logr`

Reply:
I don't think that fatal should be a log level. The SDK shouldn't be causing an exit. Instead, it should return errors so an application using the SDK can cleanly handle problems. The Helm CLI uses cobra which has a mechanism to cleanly exit (with a code). That should be used. I don't see a place where `os.Exit` should be used and have it provide a clean experience.